File: solutions/python/robot-simulator/1/robot_simulator.py
# Globals for the directions
NORTH = 0
EAST = 1
SOUTH = 2
WEST = 3


class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, commands):
        for cmd in commands:
            if cmd == 'R':
                self.direction = (self.direction + 1) % 4
            elif cmd == 'L':
                self.direction = (self.direction - 1) % 4
            elif cmd == 'A':
                x, y = self.coordinates
                if self.direction == NORTH:
                    y += 1
                elif self.direction == EAST:
                    x += 1
                elif self.direction == SOUTH:
                    y -= 1
                elif self.direction == WEST:
                    x -= 1
                self.coordinates = (x, y)

File: solutions/ruby/robot-simulator/1/robot_simulator.rb
class Robot
  attr_reader :bearing, :coordinates

  DIRECTIONS = [:north, :east, :south, :west]

  def orient(direction)
    raise ArgumentError unless DIRECTIONS.include?(direction)
    @bearing = direction
  end

  def at(x, y)
    @coordinates = [x, y]
  end

  def turn_right
    i = DIRECTIONS.index(@bearing)
    @bearing = DIRECTIONS[(i + 1) % 4]
  end

  def turn_left
    i = DIRECTIONS.index(@bearing)
    @bearing = DIRECTIONS[(i - 1) % 4]
  end

  def advance
    x, y = @coordinates
    case @bearing
    when :north then @coordinates = [x, y + 1]
    when :east  then @coordinates = [x + 1, y]
    when :south then @coordinates = [x, y - 1]
    when :west  then @coordinates = [x - 1, y]
    end
  end
end

class Simulator
  COMMANDS = { 'L' => :turn_left, 'R' => :turn_right, 'A' => :advance }

  def instructions(str)
    str.chars.map { |c| COMMANDS[c] }
  end

  def place(robot, x:, y:, direction:)
    robot.at(x, y)
    robot.orient(direction)
  end

  def evaluate(robot, str)
    instructions(str).each { |cmd| robot.send(cmd) }
  end
end
